File: src-tauri/src/model/trace.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Trace {
    pub trace_id: String,
    pub resource_spans: Vec<ResourceSpans>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ResourceSpans {
    pub resource: Resource,
    pub scope_spans: Vec<ScopeSpans>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Resource {
    pub attributes: Vec<KeyValue>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ScopeSpans {
    pub scope: InstrumentationScope,
    pub spans: Vec<Span>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct InstrumentationScope {
    pub name: String,
    #[serde(default)]
    pub version: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Span {
    pub trace_id: String,
    pub span_id: String,
    #[serde(default)]
    pub parent_span_id: Option<String>,
    pub name: String,
    pub start_time_unix_nano: String,
    pub end_time_unix_nano: String,
    #[serde(default)]
    pub attributes: Vec<KeyValue>,
    #[serde(default)]
    pub status: SpanStatus,
    #[serde(default)]
    pub events: Vec<SpanEvent>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
pub struct SpanStatus {
    #[serde(default)]
    pub code: String,
    #[serde(default)]
    pub message: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SpanEvent {
    pub time_unix_nano: String,
    pub name: String,
    #[serde(default)]
    pub attributes: Vec<KeyValue>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct KeyValue {
    pub key: String,
    pub value: AnyValue,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(untagged)]
pub enum AnyValue {
    StringValue {
        #[serde(rename = "stringValue")]
        string_value: String,
    },
    BoolValue {
        #[serde(rename = "boolValue")]
        bool_value: bool,
    },
    IntValue {
        #[serde(rename = "intValue")]
        int_value: i64,
    },
    DoubleValue {
        #[serde(rename = "doubleValue")]
        double_value: f64,
    },
    ArrayValue {
        #[serde(rename = "arrayValue")]
        array_value: Vec<AnyValue>,
    },
    KvlistValue {
        #[serde(rename = "kvlistValue")]
        kvlist_value: Vec<KeyValue>,
    },
}

File: src/lib/types/trace.ts
export interface KeyValue {
  key: string;
  value: AnyValue;
}

export type AnyValue =
  | { string_value: string }
  | { bool_value: boolean }
  | { int_value: number }
  | { double_value: number }
  | { array_value: AnyValue[] }
  | { kvlist_value: KeyValue[] };

export interface SpanEvent {
  time_unix_nano: number;
  name: string;
  attributes: KeyValue[];
}

export interface LogEntryView {
  timestamp_ns: number;
  message: string;
  labels: KeyValue[];
  span_id: string | null;
}

export interface SpanView {
  span_id: string;
  parent_span_id: string | null;
  service_name: string;
  operation_name: string;
  start_time_unix_nano: number;
  duration_ns: number;
  relative_start_ns: number;
  relative_start_pct: number;
  duration_pct: number;
  depth: number;
  has_error: boolean;
  attributes: KeyValue[];
  resource_attributes: KeyValue[];
  events: SpanEvent[];
  logs: LogEntryView[];
}

export interface TraceView {
  trace_id: string;
  root_service_name: string;
  root_operation_name: string;
  start_time_unix_nano: number;
  duration_ns: number;
  service_count: number;
  span_count: number;
  error_count: number;
  spans: SpanView[];
}
